src/passes/I64ToI32Lowering.cpp

```diff
@@ -0,0 +1,1201 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Lowers i64s to i32s by splitting variables and arguments
+// into pairs of i32s. i64 return values are lowered by
+// returning the low half and storing the high half into a
+// global.
+//
+
+#include <algorithm>
+#include "wasm.h"
+#include "pass.h"
+#include "emscripten-optimizer/istring.h"
+#include "support/name.h"
+#include "wasm-builder.h"
+
+
+namespace wasm {
+
+static Name makeHighName(Name n) {
+  return Name(
+    cashew::IString((std::string(n.c_str()) + "$hi").c_str(), false)
+  );
+}
+
+struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
+  static Name highBitsGlobal;
+
+  // false since function types need to be lowered
+  // TODO: allow module-level transformations in parallel passes
+  bool isFunctionParallel() override { return false; }
+
+  Pass* create() override {
+    return new I64ToI32Lowering;
+  }
+
+  void doWalkModule(Module* module) {
+    auto* moduleBuilder = builder = new Builder(*module);
+    PostWalker<I64ToI32Lowering>::doWalkModule(module);
+    delete moduleBuilder;
+  }
+
+  void visitFunctionType(FunctionType* curr) {
+    std::vector<WasmType> params;
+    for (auto t : curr->params) {
+      if (t == i64) {
+        params.push_back(i32);
+        params.push_back(i32);
+      } else {
+        params.push_back(t);
+      }
+    }
+    std::swap(params, curr->params);
+    if (curr->result == i64) {
+      curr->result = i32;
+    }
+  }
+
+  void visitGlobal(Global* curr) {
+    if (curr->type == i64) {
+      builder = new Builder(*getModule());
+      curr->type = i32;
+      auto* high = new Global(*curr);
+      high->name = makeHighName(curr->name);
+      getModule()->addGlobal(high);
+      delete builder;
+    }
+  }
+
+  void doWalkFunction(Function* func) {
+    // create builder here because this may be first entry to module if parallel
+    builder = new Builder(*getModule());
```

Comment:
Does this not leak?
If this *may* be the first entry, should we only create after checking that we need to?
If we're running this in parallel, is this racy either way?